File: filon-web/lib/currency.ts
// Version alignée sur filon-currency-roster-2026-08-29-v1 côté API.
// Une nouvelle devise doit être ajoutée explicitement aux deux frontières :
// trois lettres seules ne constituent pas une preuve de comparabilité.
export const SUPPORTED_CURRENCY_CODES = new Set([
  "EUR", "CHF", "GBP", "DKK", "SEK", "NOK", "ISK", "PLN", "CZK", "HUF", "RON", "BGN",
  "ALL", "BAM", "MKD", "RSD", "MDL", "UAH", "TRY", "GEL", "AMD", "AZN", "USD", "CAD",
  "AUD", "NZD", "JPY", "CNY", "HKD", "SGD", "KRW", "INR", "AED", "SAR", "ILS", "ZAR",
]);

export function normalizeSupportedCurrency(value: unknown): string | null {
  if (typeof value !== "string") return null;
  const code = value.trim().toUpperCase();
  return SUPPORTED_CURRENCY_CODES.has(code) ? code : null;
}

export type SupportedMoney = {
  amount: number;
  currency: string;
};

/** Un montant n'est affichable que si valeur et devise sont toutes deux prouvées. */
export function normalizeSupportedMoney(amount: unknown, currency: unknown): SupportedMoney | null {
  if (typeof amount !== "number" || !Number.isFinite(amount) || amount <= 0) return null;
  const code = normalizeSupportedCurrency(currency);
  return code === null ? null : { amount, currency: code };
}

/** Formate sans symbole ni devise de secours ; l'inconnue reste `null`. */
export function formatSupportedMoney(
  amount: unknown,
  currency: unknown,
  locale: "fr" | "nl" | "en" = "fr",
): string | null {
  const money = normalizeSupportedMoney(amount, currency);
  if (money === null) return null;
  const numberLocale = locale === "nl" ? "nl-BE" : locale === "en" ? "en-GB" : "fr-BE";
  try {
    return new Intl.NumberFormat(numberLocale, {
      style: "currency",
      currency: money.currency,
      maximumFractionDigits: 2,
    }).format(money.amount);
  } catch {
    return null;
  }
}

File: filon-backend/app/services/currency.py
"""Normalisation prudente des devises observées dans les flux marchands."""

from __future__ import annotations


# Roster monétaire explicitement supporté par les parcours FILON Europe et les
# flux internationaux actuellement acceptés. Il est volontairement fermé : un
# nouveau code ISO 4217 doit être examiné puis ajouté avec une nouvelle version,
# plutôt qu'être rendu comparable parce qu'il comporte simplement trois lettres.
SUPPORTED_CURRENCY_ROSTER_VERSION = "filon-currency-roster-2026-08-29-v1"
SUPPORTED_CURRENCY_CODES = frozenset(
    {
        # Marchés FILON et Europe élargie.
        "EUR",
        "CHF",
        "GBP",
        "DKK",
        "SEK",
        "NOK",
        "ISK",
        "PLN",
        "CZK",
        "HUF",
        "RON",
        "BGN",
        "ALL",
        "BAM",
        "MKD",
        "RSD",
        "MDL",
        "UAH",
        "TRY",
        "GEL",
        "AMD",
        "AZN",
        # Devises internationales rencontrées dans les catalogues partenaires.
        "USD",
        "CAD",
        "AUD",
        "NZD",
        "JPY",
        "CNY",
        "HKD",
        "SGD",
        "KRW",
        "INR",
        "AED",
        "SAR",
        "ILS",
        "ZAR",
    }
)


def normalize_currency_code(value: object) -> str | None:
    """Retourne un code monétaire exploitable, sinon l'inconnue explicite.

    Le catalogue reçoit parfois ``None``, une chaîne vide ou un libellé tel que
    ``unknown``. Aucun de ces cas ne permet de comparer un prix ni d'appliquer
    un budget. Les codes valides restent normalisés en majuscules sans inventer
    de devise à partir du pays ou du marchand.
    """

    if not isinstance(value, str):
        return None
    code = value.strip().upper()
    if code not in SUPPORTED_CURRENCY_CODES:
        return None
    return code
